common/views/components/GifVideo/GifVideo.js

```diff
@@ -0,0 +1,151 @@
+// @flow
+import React, {Component} from 'react';
+import {font, classNames} from '../../../utils/classnames';
+import {trackGaEvent} from '../../../utils/tracking';
+import debounce from 'lodash.debounce';
+import throttle from 'lodash.throttle';
+import Tasl from '../Tasl/Tasl';
+import Caption from '../Caption/Caption';
+import type {HTMLString} from '../../../services/prismic/types';
+import type {Tasl as TaslType} from '../../../model/tasl';
+type Props = {|
+  playbackRate: number,
+  videoUrl: string,
+  caption: ?HTMLString,
+  tasl: ?TaslType
+|}
+
+type State = {|
+  canPlay: boolean,
+  isPlaying: boolean,
+  autoPlayDisabled: boolean
+|}
+
+class GifVideo extends Component<Props, State> {
+  state = {
+    canPlay: false,
+    isPlaying: false,
+    autoPlayDisabled: false
+  }
+
+  videoRef = React.createRef();
+
+  inViewport = (video: HTMLElement) => {
+    const rect = video.getBoundingClientRect();
+    return (
+      rect.top >= 0 - rect.height &&
+      rect.top <= window.innerHeight &&
+      rect.left >= 0 &&
+      rect.left < window.innerWidth
+    );
+  }
+
+    playVideo = (video: HTMLMediaElement) => {
+      this.setState({
+        isPlaying: true
+      });
+      video.play();
+    }
+
+    pauseVideo = (video: HTMLMediaElement) => {
+      this.setState({
+        isPlaying: false
+      });
+      video.pause();
+    }
+
+    autoControlGif = () => {
+      const video = this.videoRef.current;
+
+      const inViewAndPlayable = video && this.inViewport(video) && this.state.canPlay;
+      if (video) {
+        if (!this.state.autoPlayDisabled) {
+          if (inViewAndPlayable) {
+            this.playVideo(video);
+          } else {
+            this.pauseVideo(video);
+          }
+        }
+      }
+    };
+
+    manualControlGif = () => {
+      const video = this.videoRef.current;
+      if (video) {
+        if (!this.state.isPlaying) {
+          this.setState({
+            autoPlayDisabled: false
+          });
+          this.playVideo(video);
+        } else {
+          this.setState({
+            autoPlayDisabled: true
+          });
+          this.pauseVideo(video);
+        }
+      }
+      trackGaEvent({
+        category: 'component',
+        action: 'toggle-gif-video-play:click',
+        label: `gif-video:${this.props.videoUrl}, click-action:${!this.state.isPlaying ? 'did-play' : 'did-pause'}`
+      });
+    };
+
+    initVideoGif = (video: HTMLMediaElement) => {
+      this.setState({
+        canPlay: true
+      });
+      setTimeout(() => { this.autoControlGif(); }, 0); // TODO - fix properly - canPlay is still false without setTimeout
```

Comment:
@jamesgorrie this is the weird bit I was talking about, without the setTimeout the value of this.state.canPlay is false when accessed in this.autoControlGIf - any ideas?